src/lib/dependencies/index.js

```diff
@@ -1,73 +1,128 @@
+  npm,
+  composer,
+  nuget,
+  dep,
+};
+
+const languageToFileType = {
+  JavaScript: 'npm',
+  TypeScript: 'npm',
+  PHP: 'composer',
+  'C#': 'nuget',
+  'Go': 'dep',
+};
+
+// Assumes dependencyObject is {core: ['dependency', ...], ...}
+// or {core: {dependency: ...}, ...}
+// Returns [{type: 'npm', name: 'dependency', core: 1}, ...]
+function transformToStats (manager, ...dependencyObjects) {
+  const index = {};
+  dependencyObjects.forEach( dependencies => {
+    Object.entries(dependencies).forEach(([ dependencyType, dependencyNames ]) => {
+      if (dependencyNames) {
+        if (!Array.isArray(dependencyNames)) { // Transform {dependency: ...} into ['dependency', ...]
+          dependencyNames = Object.keys(dependencyNames);
+        }
+        dependencyNames.forEach(name => {
+          index[name] = index[name] || { type: manager, name };
+          index[name][dependencyType] = 1;
+        });
+      }
+    });
+  });
+  return Object.values(index);
+}
+
 function getDependenciesFromGithubRepo (githubRepo, githubAccessToken) {
-  const strategies = [];
-  if (githubRepo.language === 'JavaScript' || githubRepo.language === 'TypeScript') {
-    strategies.push(npm.getDependenciesFromGithubRepo(githubRepo, githubAccessToken));
-  }
-  if (githubRepo.language === 'PHP') {
-    strategies.push(composer.getDependenciesFromGithubRepo(githubRepo, githubAccessToken));
-  }
-  if (githubRepo.language === 'C#') {
-    strategies.push(nuget.getDependenciesFromGithubRepo(githubRepo, githubAccessToken));
-  }
-  if (githubRepo.language === 'Go') {
-    strategies.push(dep.getDependenciesFromGithubRepo(githubRepo, githubAccessToken));
+  const fileType = languageToFileType[githubRepo.language];
+  if (!fileType) {
+    return Promise.resolve([]);
+  }
+  const cacheKey = `repo_${fileType}_dependencies_${githubRepo.id}`;
+  if (cache.has(cacheKey)) {
+    return cache.get(cacheKey);
+  }
+  return loadDependenciesFromGithubRepo (fileType, githubRepo, githubAccessToken)
+    .then(result => {
+      cache.set(cacheKey, result);
+      return result;
+    });
+}
+
+function loadDependenciesFromGithubRepo (fileType, githubRepo, githubAccessToken) {
+  const manager = dependencyManagers[fileType];
+  const query = manager.searchAllRepo ? searchFilesFromRepo :
+    (repo, path, accessToken) => fetchFileFromRepo (repo, path, accessToken).then( content => [content] );
+
+  // lookup returns a promise returning [{matchedPattern, content}]
+  function lookup (pattern) {
+    return query(githubRepo, pattern, githubAccessToken).then (
+      contents => contents.map(text => ({
+              matchedPattern: pattern,
+              text,
+            }))
+    );
   }
-  return Promise.all(strategies).then(results => {
-    return flatten(results, true);
+
+  const firstSetOfMatchingFiles = manager.patterns.reduce(
+    (promise, pattern) => promise.then (
+      results => results.length ? results : lookup(pattern)
+    ), Promise.resolve([])
+  );
```

Comment:
From line 68 to 84, I find the code difficult to read (mix of closures, promises, reduce). If that was possible, I would love to see something simpler.

Reply:
Doubling on that, at the moment, this complexity is only for the `nuget` strategy.

As a general direction, if something complex is not shared between multiple strategies, then I prefer if the complexity stay in the strategy itself.

Reply:
Let me try making code simpler with `async/await`.

>  if something complex is not shared between multiple strategies, then I prefer if the complexity stay in the strategy itself.

It's always tough to predict the future, but the direction I took was to minimize the burden for implementors of package managers. The resulting files are very simple and they do the minimum, which is interpret the file for dependencies and project name, as well as give the naming scheme (`patterns`).

I feel like everything else does not belong in individual package manager source. There could be other communities that don't put their dependency files at the fixed place. The global strategy could also be subject to change; say there's no more drastic query limits, then maybe we'll want to change the strategy to "check root for the right files, if none found, do a query in the whole repo" (even for npm/composer/...). IMO, that is not a concern of each individual manager.